sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
         BinImages.push_back(ImageUPtr.get());
     }
   }
+  // TODO: Add a diagnostic on multiple device images with conflicting kernel
+  // names.
 
+  // TODO: Cache device_image objects
   // Create SYCL device image from those that have compatible state and at least
   // one device
   std::vector<device_image_plain> SYCLDeviceImages;
   for (RTDeviceBinaryImage *BinImage : BinImages) {
     const bundle_state ImgState = getBinImageState(BinImage);
-    // Ignore images with incompatible state. Image is considered compatible
-    // with a target state if an image is already in the target state or can be
-    // brought to target state by compiling/linking/building.
-    //
-    // Example: an image in "executable" state is not compatbile with "input"
-    // target state - there is no operation to convert the image it to "input"
-    // state.
-    // An image in "input" state is compatible with "executable" target state
-    // because it can be built to get into "executable" state.
-    if (ImgState > TargetState)
+    if (ImgState != TargetState)
       continue;
 
     for (const sycl::device &Dev : Devs)
       if (compatibleWithDevice(BinImage, Dev)) {
+
+        // TODO: Cache kernel_ids
+        std::vector<sycl::kernel_id> KernelIDs;
+        // Collect kernel names for the image
+        pi_device_binary DevBin =
+            const_cast<pi_device_binary>(&BinImage->getRawData());
+        for (_pi_offload_entry EntriesIt = DevBin->EntriesBegin;
+             EntriesIt != DevBin->EntriesEnd; ++EntriesIt) {
+
+          std::shared_ptr<detail::kernel_id_impl> KernelIDImpl =
+              std::make_shared<detail::kernel_id_impl>(EntriesIt->name);
+
+          KernelIDs.push_back(
+              detail::createSyclObjFromImpl<sycl::kernel_id>(KernelIDImpl));
+        }
+        // device_image_impl expects kernel ids to be sorted for fast search
+        std::sort(KernelIDs.begin(), KernelIDs.end(), LessByNameComp{});
+
         DeviceImageImplPtr Impl = std::make_shared<detail::device_image_impl>(
-            BinImage, Ctx, Devs, ImgState);
+            BinImage, Ctx, Devs, ImgState, KernelIDs, /*PIProgram=*/nullptr);
 
         SYCLDeviceImages.push_back(
             createSyclObjFromImpl<device_image_plain>(Impl));
         break;
       }
   }
 
-  // Make it so that SYCL device images have required state by compiling or
-  // linking them if needed
-  switch (TargetState) {
-  case bundle_state::input:
-    // Do nothing as the check above should make sure that resulting images are
-    // in input state already
-    break;
-  case bundle_state::object:
-    assert(false && "Not implemented yet");
-    break;
-  case bundle_state::executable:
-    assert(false && "Not implemented yet");
-    break;
+  return SYCLDeviceImages;
+}
+
+device_image_plain
+ProgramManager::compile(const device_image_plain &DeviceImage,
+                        const std::vector<device> &Devs,
+                        const property_list &) {
```

Comment:
I don't see any way to pass options to compilation or linkage below. Is this a TODO? please add if so

Reply:
It is not defined in the spec, but it looks like intention is to pass them using property_list.
Not sure which TODO is needed here.

Reply:
So, the `property_list` parameter should serve this purpose, now I get it. But it is not used and I can see `/*options=*/nullptr`. So I'd suggest to add something like:
```cpp
// TODO Add translation of the property_list parameter to compilation options and pass them to `piProgramCompile`
```

Reply:
If we want to support any compilation options passed via the property list, we would need to update the Level Zero backend specification (or perhaps some other extension spec) to document what those options are.  Until we do that, there is nothing TODO in the code here.  :-)

Reply:
@gmlueck Could you please clarify how compile and link options can be passed in SYCL2020 API?

Reply:
The definitions of `compile()`, `link()`, and `build()` all take a `property_list` parameter, but the core spec doesn't define any properties.  Instead, this is an extension point for vendors and backends.  Let's suppose that we wanted to expose the "-ze-opt-large-register-file" build option from Level Zero.  We would specify an extension that defines a new property like:

```
namespace sycl::ext::oneapi::level_zero {

namespace property {
struct large_register_file {};
// namespace property

static constexpr property::large_register_file large_register_file;

} // namespace sycl::ext::oneapi::level_zero
```

Applications could then build with this option like this:

```
build(inputBundle, {ext::oneapi::level_zero::large_register_file});
```

We could decide whether to expose this as a DPC++ extension or as part of the Level Zero Backend spec.  Assuming that these compilation options are considered part of the Level Zero API, it probably makes sense to expose the SYCL integration as part of the Level Zero backend spec.

Reply:
Added a comment.